pkg/rebalancer/storage/controller/service/rebalancer.go

```diff
@@ -19,28 +19,544 @@ package service
+// And it returns the rebalancer implemenation or any error occurred.
+func NewRebalancer(opts ...RebalancerOption) (Rebalancer, error) {
 	r := new(rebalancer)
-	for _, opt := range append(defaultOpts, opts...) {
+
+	for _, opt := range append(defaultRebalancerOpts, opts...) {
 		if err := opt(r); err != nil {
 			return nil, errors.ErrOptionFailed(err, reflect.ValueOf(opt))
 		}
 	}
+
+	err := r.initCtrl()
+	if err != nil {
+		return nil, err
+	}
+
+	r.decoder, err = decoder.NewDecoder(r.ctrl.GetManager().GetScheme())
+	if err != nil {
+		return nil, err
+	}
+
+	r.jobObject = new(job.Job)
+	err = r.decoder.DecodeInto(*(*[]byte)(unsafe.Pointer(&r.jobTemplate)), r.jobObject)
+	if err != nil {
+		return nil, errors.ErrFailedToDecodeJobTemplate(err)
+	}
+
 	return r, nil
 }
 
+func (r *rebalancer) initCtrl() (err error) {
+	job, err := job.New(
+		job.WithControllerName("job rebalancer"),
+		job.WithNamespaces(r.jobNamespace),
+		job.WithOnErrorFunc(func(err error) {
+			log.Error(err)
+		}),
+		job.WithOnReconcileFunc(r.jobReconcile),
+	)
+	if err != nil {
+		return err
+	}
+
+	var rc k8s.ResourceController
+	switch r.agentResourceType {
+	case config.STATEFULSET:
+		rc, err = statefulset.New(
+			statefulset.WithControllerName("statefulset rebalancer"),
+			statefulset.WithNamespaces(r.agentNamespace),
+			statefulset.WithOnErrorFunc(func(err error) {
+				log.Error(err)
+			}),
+			statefulset.WithOnReconcileFunc(r.statefulsetReconcile),
+		)
+		if err != nil {
+			return err
+		}
+	case config.REPLICASET:
+		// TODO: implment get replicaset reconciled result
+		return nil
+	case config.DAEMONSET:
+		// TODO: implment get daemonset reconciled result
+		return nil
+	default:
+		return errors.ErrInvalidAgentResourceType(r.agentResourceType.String())
+	}
+
+	r.ctrl, err = k8s.New(
+		k8s.WithControllerName("rebalance storage controller"),
+		k8s.WithEnableLeaderElection(),
+		k8s.WithLeaderElectionID(r.leaderElectionID),
+		k8s.WithResourceController(job),
+		k8s.WithResourceController(rc),
+		k8s.WithResourceController(pod.New(
+			pod.WithControllerName("pod discover"),
+			pod.WithOnErrorFunc(func(err error) {
+				log.Error(err)
+			}),
+			pod.WithOnReconcileFunc(r.podReconcile),
+		)),
+		k8s.WithResourceController(mpod.New(
+			mpod.WithControllerName("pod metrics discover"),
+			mpod.WithOnErrorFunc(func(err error) {
+				log.Error(err)
+			}),
+			mpod.WithOnReconcileFunc(r.podMetricsReconcile),
+		)),
+	)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (r *rebalancer) jobReconcile(jobList map[string][]job.Job) {
+	log.Debugf("[reconcile Job] length Joblist: %d", len(jobList))
+	jobs, ok := jobList[r.jobName]
+	if ok {
+		r.jobs.Store(jobs)
+		return
+	}
+
+	r.jobs.Store(make([]job.Job, 0))
+	log.Infof("job not found: %s", r.jobName)
+}
+
+func (r *rebalancer) statefulsetReconcile(ctx context.Context, statefulSetList map[string][]statefulset.StatefulSet) {
```

Comment:
@kpango @rinx you can skip reviewing this function as we may change the logic later.